sw/device/lib/dif/dif_usbdev.c

```diff
@@ -0,0 +1,1061 @@
+         .bit_index = USBDEV_IN_SENT_SENT##N}
+
+static const endpoint_hw_info_t kEndpointHwInfos[USBDEV_NUM_ENDPOINTS] = {
+    ENDPOINT_HW_INFO_ENTRY(0),  ENDPOINT_HW_INFO_ENTRY(1),
+    ENDPOINT_HW_INFO_ENTRY(2),  ENDPOINT_HW_INFO_ENTRY(3),
+    ENDPOINT_HW_INFO_ENTRY(4),  ENDPOINT_HW_INFO_ENTRY(5),
+    ENDPOINT_HW_INFO_ENTRY(6),  ENDPOINT_HW_INFO_ENTRY(7),
+    ENDPOINT_HW_INFO_ENTRY(8),  ENDPOINT_HW_INFO_ENTRY(9),
+    ENDPOINT_HW_INFO_ENTRY(10), ENDPOINT_HW_INFO_ENTRY(11),
+};
+
+#undef ENDPOINT_HW_INFO_ENTRY
+
+/**
+ * Mapping from `dif_usbdev_irq_t` to bit indices in interrupt registers.
+ */
+static const uint8_t kIrqEnumToBitIndex[] = {
+        [kDifUsbdevIrqPktReceived] = USBDEV_INTR_COMMON_PKT_RECEIVED,
+        [kDifUsbdevIrqPktSent] = USBDEV_INTR_COMMON_PKT_SENT,
+        [kDifUsbdevIrqDisconnected] = USBDEV_INTR_COMMON_DISCONNECTED,
+        [kDifUsbdevIrqHostLost] = USBDEV_INTR_COMMON_HOST_LOST,
+        [kDifUsbdevIrqLinkReset] = USBDEV_INTR_COMMON_LINK_RESET,
+        [kDifUsbdevIrqLinkSuspend] = USBDEV_INTR_COMMON_LINK_SUSPEND,
+        [kDifUsbdevIrqLinkResume] = USBDEV_INTR_COMMON_LINK_RESUME,
+        [kDifUsbdevIrqAvEmpty] = USBDEV_INTR_COMMON_AV_EMPTY,
+        [kDifUsbdevIrqRxFull] = USBDEV_INTR_COMMON_RX_FULL,
+        [kDifUsbdevIrqAvOverflow] = USBDEV_INTR_COMMON_AV_OVERFLOW,
+        [kDifUsbdevIrqLinkInError] = USBDEV_INTR_COMMON_LINK_IN_ERR,
+        [kDifUsbdevIrqRxCrcError] = USBDEV_INTR_COMMON_RX_CRC_ERR,
+        [kDifUsbdevIrqRxPidError] = USBDEV_INTR_COMMON_RX_PID_ERR,
+        [kDifUsbdevIrqRxBitstuffError] = USBDEV_INTR_COMMON_RX_BITSTUFF_ERR,
+        [kDifUsbdevIrqFrame] = USBDEV_INTR_COMMON_FRAME,
+        [kDifUsbdevIrqConnected] = USBDEV_INTR_COMMON_CONNECTED,
+};
+
+/**
+ * Static functions for the free buffer pool.
+ */
+
+/**
+ * Checks if a buffer pool is full.
+ *
+ * A buffer pool is full if it contains `USBDEV_NUM_BUFFERS` entries.
+ *
+ * @param pool A buffer pool.
+ * @return `true` if the buffer pool if full, `false` otherwise.
+ */
+static bool buffer_pool_is_full(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == BUFFER_POOL_FULL;
+}
+
+/**
+ * Checks if a buffer pool is empty.
+ *
+ * @param pool A buffer pool.
+ * @return `true` if the buffer pool is empty, `false` otherwise.
+ */
+static bool buffer_pool_is_empty(dif_usbdev_buffer_pool_t *pool) {
+  return pool->top == BUFFER_POOL_EMPTY;
+}
+
+/**
+ * Checks if a buffer entry is valid.
+ *
+ * An entry is valid if it is less than `USBDEV_NUM_BUFFERS`.
+ *
+ * @param entry A buffer entry.
+ * @return `true` if the entry is valid, `false` otherwise.
+ */
+static bool buffer_pool_is_valid_entry(uint8_t entry) {
+  return entry < USBDEV_NUM_BUFFERS;
+}
+
+/**
+ * Adds a buffer entry to a buffer pool.
+ *
+ * @param pool A buffer pool.
+ * @param entry A buffer entry.
+ * @return `true` if the operation was successful, `false` otherwise.
+ */
+static bool buffer_pool_add(dif_usbdev_buffer_pool_t *pool, uint8_t entry) {
```

Comment:
Nit: since this is supposed to behave like a stack, maybe `push` and `pop` are more descriptive names? `remove` especially suggests removing a specific index.

Reply:
I went back and forth on this before using `add` and `remove`, but my rationale was: "This is a buffer pool, which happens to be implemented as a stack, so it feels more natural to have `add` and `remove` as opposed to `push` and `pop`." Please let me know what you think.

Reply:
Fair enough. I would still consider another name than `remove`, such as `take`. I'll leave it to your judgement.